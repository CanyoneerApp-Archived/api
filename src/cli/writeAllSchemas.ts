import FS from 'fs';
import Path from 'path';
import * as TSJ from 'ts-json-schema-generator';

export const schemas = {
  RouteV1: 'v1',
  RouteV2: 'v2',
  IndexRouteV2: 'v2',
  GeoJSONRouteV2: 'v2',
};

export async function writeAllSchemas() {
  return Promise.all(
    Object.entries(schemas).map(async ([type, version]) => {
      const outputPath = Path.resolve(`./public/${version}/schemas/${type}.json`);
      await FS.promises.mkdir(Path.dirname(outputPath), {recursive: true});
      await FS.promises.writeFile(
        outputPath,
        JSON.stringify(
          TSJ.createGenerator({
            path: `./src/types/${version}.ts`,
            tsconfig: './tsconfig.json',
            topRef: false,
          }).createSchema(type),
          null,
          2,
        ),
      );
    }),
  );
}
