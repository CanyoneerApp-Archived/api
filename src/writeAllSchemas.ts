import FS from 'fs';
import * as TSJ from 'ts-json-schema-generator';

const types = {
  RouteV1: './src/scrape/RouteV1.ts',
  RouteV2: './src/scrape/RouteV2.ts',
  IndexRouteV2: './src/scrape/RouteV2.ts',
  RouteV2GeoJSONFeature: './src/scrape/RouteV2.ts',
};

export async function writeAllSchemas() {
  await FS.promises.mkdir('./output/schemas', {recursive: true});

  Object.entries(types).map(([type, path]) =>
    FS.promises.writeFile(
      `./output/schemas/${type}.json`,
      JSON.stringify(
        TSJ.createGenerator({
          path,
          tsconfig: './tsconfig.json',
          topRef: false,
        }).createSchema(type),
        null,
        2,
      ),
    ),
  );
}
