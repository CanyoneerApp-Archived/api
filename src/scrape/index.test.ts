import {inner as scrapeInner} from '.';
import {writeAllSchemas} from '../writeAllSchemas';

describe('scrape', () => {
  beforeAll(async () => {
    await writeAllSchemas();
  });

  it(
    'Maine',
    async () => {
      return expect(await scrapeInner(['Maine'])).toMatchSnapshot();
    },
    60 * 1000,
  );

  it(
    'South Dakota',
    async () => {
      return expect(await scrapeInner(['South Dakota'])).toMatchSnapshot();
    },
    60 * 1000,
  );
});
