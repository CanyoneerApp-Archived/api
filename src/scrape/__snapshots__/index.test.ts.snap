// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrape Maine 1`] = `
[
  {
    "description": "<h2 id="introduction">Introduction</h2>
<p>This gorgeous canyon sits high up in the White Mountains of western
Maine. The canyon descends over 400 feet in just a half mile and
requires technical gear for navigation around its four rappels. It may
be one of the most enjoyable descents in western Maine. Spend a fun
morning descending this canyon then hop on over to Rattlesnake Gorge for
another fun descent.</p>
<h2 id="approach">Approach</h2>
<p>The best location to start from is the Brickett Place. There is a day
use fee for this trailhead it was $5.00 in 2021. From the trailhead hike
up Bickford Brook Trail past the first trail intersection until you
reach the second fork in the trail. (19 T 0341714E 4904338N). Take the
trail on the right at this intersection. The trail will cross over the
creek, follow this creek downstream until you reach the canyon.</p>
<h2 id="descent">Descent</h2>
<p>Wrap a tree and descend 30 ft into the canyon. Continue down canyon
until you reach the second drop. Wrap another tree and descend 20 ft.
Following the second rappel hike along the canyon avoiding the pools of
water as much as possible. Turn the corner and wrap another tree or
boulder and descend to the large pool at the end of the canyon. Hike
along the stream another 1/4th of a mile to where a trail crosses the
stream.</p>
<h2 id="exit">Exit</h2>
<p>Hike out to the trail head on this trail or descend down one last
rappel on the other side of this trail.</p>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<p><a
href="https://eyesupadventure.wixsite.com/adventure/maine-canyoneering">https://eyesupadventure.wixsite.com/adventure/maine-canyoneering</a></p>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": undefined,
    "id": "68720",
    "latitude": 44.274,
    "longitude": -70.9843,
    "months": [
      "Jun",
      "Jul",
      "Aug",
      "Sep",
      "Oct",
    ],
    "name": "Bickford Slides Canyon",
    "permits": "No",
    "quality": 3,
    "rappelCountMax": 3,
    "rappelCountMin": 3,
    "rappelLongestFeet": 50,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": 3,
    "timeRating": "II",
    "url": "https://ropewiki.com/Bickford_Slides_Canyon",
    "vehicle": undefined,
    "waterRating": "c",
  },
  {
    "description": "<h2 id="introduction">Introduction</h2>
<p>This is one of the only caves in Maine that requires the use of SRT
equipment. The cave is small but makes for a fun adventure. Location
beta and more information can be provided by contacting the Maine Cave
Survey with the following email: stephen.whitney1999@gmail.com</p>
<p>If you have visited this cave please do not post location data to
keep riffraff out of dangrous situations!</p>
<h2 id="approach">Approach</h2>
<h2 id="descent">Descent</h2>
<h2 id="exit">Exit</h2>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": undefined,
    "id": "68718",
    "latitude": 44.2165,
    "longitude": -69.0747,
    "months": [],
    "name": "Inman's Cave",
    "permits": "No",
    "quality": 0,
    "rappelCountMax": 2,
    "rappelCountMin": 2,
    "rappelLongestFeet": 20,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": 3,
    "timeRating": "I",
    "url": "https://ropewiki.com/Inman%27s_Cave",
    "vehicle": undefined,
    "waterRating": "a",
  },
  {
    "description": "<h2 id="introduction">Introduction</h2>
<h2 id="approach">Approach</h2>
<h2 id="descent">Descent</h2>
<h2 id="exit">Exit</h2>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<ul>
<li><img src="rwl_en.png" title="rwl_en.png" alt="rwl_en.png" /><a
href="http://brennen.caltech.edu/swhikes/katadn.htm">Chris Brennen's
Adventure Hikes (Southwest)</a> : Katahdin; Maine</li>
</ul>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": {
      "features": [
        {
          "geometry": {
            "coordinates": [
              -116.329666,
              35.030834,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "Hike Beyond the campground and the Abol trailhead (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt1",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.936668,
              45.898499,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "At the top of the slide is a large cairn and a few yards beyond this at (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt2",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.921997,
              45.904335,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "It rises fairly gently to Baxter Peak (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt3",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.936668,
              45.898499,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "Pamola Peak the Knife Edge and South Peak from Baxter Peak Looking down the Hunt Trail ridge Beginning the descent you proceed back along the way you came to the the trail junction at Thoreau Spring at (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt4",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.981667,
              45.896168,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "9miles from the start at (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt5",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.998833,
              45.887333,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "1mile from the bridge you finally reach the Katahdin Stream Campground and Trailhead at (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt6",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.963997,
              45.873833,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "Br (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt7",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.936668,
              45.898499,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "28am 4380ftMW Cairn at top of slide - just a short way to trail junction at Thoreau Spring (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt8",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.921997,
              45.904335,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "30am 4910ftMW Summit of Baxter Peak (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt9",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.947334,
              45.900002,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "5mile to (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt10",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
        {
          "geometry": {
            "coordinates": [
              -68.981667,
              45.896168,
              0,
            ],
            "type": "Point",
          },
          "properties": {
            "description": "18pm 1430ftMW After drop through forest and along the side of a stream - beautiful Katahdin Stream Falls (Data by brennen.caltech.edu/swhikes)",
            "icon": "http://maps.google.com/mapfiles/kml/shapes/open-diamond.png",
            "name": "Wpt11",
            "styleHash": "-56beb8bb",
            "styleMapHash": {
              "highlight": "#dot1",
              "normal": "#dot0",
            },
            "styleUrl": "#dot",
          },
          "type": "Feature",
        },
      ],
      "type": "FeatureCollection",
    },
    "id": "25753",
    "latitude": 45.9043,
    "longitude": -68.922,
    "months": [],
    "name": "Mount Katahdin",
    "permits": "No",
    "quality": 5,
    "rappelCountMax": undefined,
    "rappelCountMin": undefined,
    "rappelLongestFeet": undefined,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": undefined,
    "timeRating": undefined,
    "url": "https://ropewiki.com/Mount_Katahdin",
    "vehicle": undefined,
    "waterRating": undefined,
  },
  {
    "description": "<h2 id="introduction">Introduction</h2>
<p>"The last canyon along the Bear River in Grafton Notch is the Screw
Auger Gorge. This short Gorge is probably one of the most popular
tourist destinations in the area. It is frequented by thousands of
tourists each summer and it is not uncommon to find families and teens
swimming here on a warm summer afternoon. The canyoneering opportunities
are limited but still possible. Rappel off a natural anchor or tree into
the pothole below, escape the small pothole and hike/ swim through the
rest of the gorge."</p>
<p>from <em>The Maine Canyon Survey: A Guide to Canyoneering in
Maine.</em></p>
<p><a
href="https://eyesupadventure.wixsite.com/adventure/maine-canyoneering">1</a></p>
<h2 id="approach">Approach</h2>
<h2 id="descent">Descent</h2>
<h2 id="exit">Exit</h2>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<ul>
<li><img src="rwl_en.png" title="rwl_en.png" alt="rwl_en.png" /><a
href="http://summitpost.org/screw-auger-falls-gorge/826303">SummitPost.org</a>
: Screw Auger Falls Gorge]</li>
</ul>
<ul>
<li><img src="rwl_en.png" title="rwl_en.png" alt="rwl_en.png" /><a
href="https://eyesupadventure.wixsite.com/adventure/maine-canyoneering">Eyes
Up Adventure</a> : Maine canyoneering]</li>
</ul>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": undefined,
    "id": "34280",
    "latitude": 44.571944444444,
    "longitude": -70.91,
    "months": [
      "Jun",
      "Jul",
      "Aug",
      "Sep",
      "Oct",
    ],
    "name": "Screw Auger Falls Gorge",
    "permits": "No",
    "quality": 0,
    "rappelCountMax": 1,
    "rappelCountMin": 1,
    "rappelLongestFeet": 25,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": 3,
    "timeRating": "I",
    "url": "https://ropewiki.com/Screw_Auger_Falls_Gorge",
    "vehicle": undefined,
    "waterRating": "c1",
  },
  {
    "description": "<h2 id="introduction">Introduction</h2>
<p>Nice Canyon near Andover, ME.</p>
<h2 id="approach">Approach</h2>
<p>Hike up trail to slot.</p>
<h2 id="descent">Descent</h2>
<p>8ft rap into slot, 25ft two tier rap at the end of slot. Another 30
ft rap and a final 50ft.</p>
<h2 id="exit">Exit</h2>
<p>Hike out trail</p>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<p>See The Maine Canyon Survey</p>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": undefined,
    "id": "68931",
    "latitude": 44.6351,
    "longitude": -70.8616,
    "months": [],
    "name": "The Cataracts",
    "permits": "No",
    "quality": 3,
    "rappelCountMax": 5,
    "rappelCountMin": 4,
    "rappelLongestFeet": 50,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": 3,
    "timeRating": "II",
    "url": "https://ropewiki.com/The_Cataracts",
    "vehicle": undefined,
    "waterRating": "c",
  },
]
`;

exports[`scrape South Dakota 1`] = `
[
  {
    "description": "<h2 id="introduction">Introduction</h2>
<h2 id="approach">Approach</h2>
<h2 id="descent">Descent</h2>
<h2 id="exit">Exit</h2>
<h2 id="red_tape">Red tape</h2>
<h2 id="beta_sites">Beta sites</h2>
<ul>
<li><img src="rwl_en.png" title="rwl_en.png" alt="rwl_en.png" /><a
href="http://summitpost.org/beaver-creek-canyon/972748">SummitPost.org</a>
: Beaver Creek Canyon</li>
</ul>
<h2 id="trip_reports_and_media">Trip reports and media</h2>
<h2 id="background">Background</h2>
",
    "geojson": undefined,
    "id": "34288",
    "latitude": 43.5815,
    "longitude": -103.4837,
    "months": [],
    "name": "Beaver Creek Canyon",
    "permits": "No",
    "quality": 0,
    "rappelCountMax": undefined,
    "rappelCountMin": undefined,
    "rappelLongestFeet": undefined,
    "riskRating": undefined,
    "shuttleMinutes": undefined,
    "technicalRating": undefined,
    "timeRating": undefined,
    "url": "https://ropewiki.com/Beaver_Creek_Canyon",
    "vehicle": undefined,
    "waterRating": undefined,
  },
]
`;
